<img src="images/logo.jpg" style="display: block; margin-left: auto; margin-right: auto;" alt="לוגו של מיזם לימוד הפייתון. נחש מצויר בצבעי צהוב וכחול, הנע בין האותיות של שם הקורס: לומדים פייתון. הסלוגן המופיע מעל לשם הקורס הוא מיזם חינמי ללימוד תכנות בעברית.">

# <p style="text-align:right;direction:rtl;"> תרגילים </p> 

## <p style="text-align:right;direction:rtl;"> בחירות במדינת הגמדים </p>

<p style="text-align:right;direction:rtl;">
    במדינת הגמדים רעש מהומה, ולכן החליטו ללכת לבחירות.<br>
    ועדת הבחירות המרכזית, בראשות דרדסבא, ערכה בחירות חשאיות והוגנות והציגה לפניכם את התוצאות בקובץ elections.txt המצורף (בתיקיית resources), על כל 353,400 בוחריה.<br>
    הקובץ הינו רשימה של שמות המפלגות שבהן בחרו תושבי מדינת הגמדים, כאשר כל שורה מייצגת בחירה של תושב מדינת הגמדים.<br>
    כדי לשמור על טוהר הבחירות לא נאמר לכם במפורש מהו מספר המפלגות שהגישו את רשימותיהן לוועדת הבחירות.<br>
    עליכם הוטלה המשימה לעבור על קובץ תוצאות הבחירות ולדווח כמה קולות קיבלה כל מפלגה.<br>
    כתבו פונקציה שלא מקבלת פרמטרים ומחזירה את תוצאות הבחירות במילון ובו שם המפלגה ומספר הקולות שקיבלה.
</p>

In [80]:
#upload 67
def get_votes_list():

    with open('resources/elections.txt' , 'r') as elections_file:

        votes_list = elections_file.read()
        votes_list = votes_list.split("\n")
        return votes_list

def get_votes_dict(votes_list):

    # putting all the kind of parties in the dict:

    votes_dict = {}
    party_votes_count = 0

    for party_name in votes_list:

        if party_name not in votes_dict:

            party_votes_count = votes_list.count(party_name)
            votes_dict[party_name] = party_votes_count

    print(votes_dict)
    return votes_dict

def passed_block_precentage(votes_dict):

    passed_dict = {}
    didnt_passed_dict = {}

    for party_name, party_result in votes_dict.items():

        if votes_dict[party_name] >= 35340:
            passed_dict[party_name] = party_result
        else:
            didnt_passed_dict[party_name] = party_result
    print("passed the block precentage:\n")
    print(passed_dict)
    print()
    print("didn't passed the block precentage:\n")       
    print(didnt_passed_dict)

def get_votes_for_passed_parties(votes_dict):

    passed_dict = {}
    didnt_passed_dict = {}
    passed_parties_sum_of_votes = 0

    for party_name, party_result in votes_dict.items():

        if votes_dict[party_name] >= 35340:
            passed_dict[party_name] = party_result
            passed_parties_sum_of_votes += party_result

    print(f"The sum of votes for the passed parties is : {passed_parties_sum_of_votes}.")

    return passed_dict

def get_mandate_dict(votes_for_passed_parties_dict , votes_per_mandate):

    mandate_dict = {}

    for party_name, party_votes in votes_for_passed_parties_dict.items():

        mandate_dict[party_name] = round(party_votes / votes_per_mandate)

    print(mandate_dict)

    return mandate_dict

def get_votes_per_mandate_ratio(votes_per_mandate_dict):

    print("im in votes per mandate ratio")
    sum_of_votes = 0

    for party_name, party_votes in votes_per_mandate_dict.items():

        sum_of_votes += party_votes

    votes_per_mandate = round(sum_of_votes / 120)

    return votes_per_mandate

def print_possible_coalition(mandate_dict):

    MANDATE_FOR_COALITION = 61

    possible_coalition_for_NilsIsALeader = []
    possible_coalition_for_MidgeLandIsUs = []

    NilsIsALeader_mandates = mandate_dict['NilsIsALeader']
    MidgeLandIsUs_mandates = mandate_dict['MidgeLandIsUs']
    
    for party_name , certain_party_mandetes in mandate_dict.items():

    # possible coalition for NilsIsALeader:
       if party_name != 'MidgeLandIsUs' and party_name != 'NilsIsALeader':

            if certain_party_mandetes + NilsIsALeader_mandates >= 61:
                 possible_coalition_for_NilsIsALeader.append(party_name)

            if  certain_party_mandetes + MidgeLandIsUs_mandates >= 61:
                 possible_coalition_for_MidgeLandIsUs.append(party_name)

    print('\n') 
    print(f"possible coalition for NilsIsALeader is:")

    if len(possible_coalition_for_NilsIsALeader) == 0:
        print("There are no possible coalition for possible_coalition for NilsIsALeader.\n")

    else:
        for party_name in possible_coalition_for_NilsIsALeader:

            print(party_name)

    print(f"possible coalition for MidgeLandIsUs is:")

    if len(possible_coalition_for_MidgeLandIsUs) == 0:
        print("There are no possible coalition for possible coalition for NilsIsALeader MidgeLandIsUs.\n")

    else:
        for party_name in possible_coalition_for_MidgeLandIsUs:

            print(party_name)

def get_election_results():

    votes_list = get_votes_list()

    votes_dict = get_votes_dict(votes_list)

    passed_block_precentage_dict = passed_block_precentage(votes_dict)

    votes_for_passed_parties = get_votes_for_passed_parties(votes_dict)

    votes_per_mandate = get_votes_per_mandate_ratio(votes_for_passed_parties)

    print(f"votes per mandate is {votes_per_mandate}.")

    mandate_dict = get_mandate_dict(votes_for_passed_parties ,votes_per_mandate)

    print_possible_coalition(mandate_dict)
get_election_results()

{'OnlyTovTov': 41075, 'NilsIsALeader': 63702, 'DorothyForMidgeLand': 23071, 'MidgeLandIsUs': 68582, 'SnufkinForCouncil': 35830, 'WiliWanka': 26628, 'GulliverIsBest': 25928, 'BestForMidgeLand': 40144, 'FingeroniBringsPeace': 22457, 'GreenMidgeLand': 5983}
passed the block precentage:

{'OnlyTovTov': 41075, 'NilsIsALeader': 63702, 'MidgeLandIsUs': 68582, 'SnufkinForCouncil': 35830, 'BestForMidgeLand': 40144}

didn't passed the block precentage:

{'DorothyForMidgeLand': 23071, 'WiliWanka': 26628, 'GulliverIsBest': 25928, 'FingeroniBringsPeace': 22457, 'GreenMidgeLand': 5983}
The sum of votes for the passed parties is : 249333.
im in votes per mandate ratio
votes per mandate is 2078.
{'OnlyTovTov': 20, 'NilsIsALeader': 31, 'MidgeLandIsUs': 33, 'SnufkinForCouncil': 17, 'BestForMidgeLand': 19}


possible coalition for NilsIsALeader is:
There are no possible coalition for possible_coalition for NilsIsALeader.

possible coalition for MidgeLandIsUs is:
There are no possible coalition for possib

<p style="text-align:right;direction:rtl;">
    אחוז החסימה במדינת הגמדים הוא אכזרי למדי ועומד על 10%.<br>
    כיון שהצביעו 353,400 גמדים – אחוז החסימה עומד על
    $353,400*0.1=35,340$ קולות.<br>
    כתבו פונקציה המקבלת את מילון תוצאות הבחירות ומדפיסה אילו מפלגות עברו את אחוז החסימה ואילו לא.
</p>

In [ ]:

<p style="text-align:right;direction:rtl;">
כעת כתבו פונקציה המקבלת את מילון תוצאות הבחירות ומחזירה את מספר הקולות שקיבלו רק מפלגות שעברו את אחוז החסימה.</p>

<p style="text-align:right;direction:rtl;">
    אחרי שסיימתם בדקו שקיבלתם את הערך הבא: <span style="direction: rtl; background: #000; text: #000">249,333</span>
</p>

<p style="text-align:right;direction:rtl;">
    לאחר קבלת תוצאות האמת המפלגות פועלות להקים קואליציה. קואליציה מורכבת מ־61 מנדטים לפחות.<br>
    מנדט שווה: $\dfrac{votes}{120}=2.077.775\cong2,078$ קולות.<br>
    שימו לב, הקולות בתרגיל זה הם התוצאה של הסעיף הקודם. כלומר הקולות בניכוי הקולות עבור מפלגות שלא עברו את אחוז החסימה.<br>
    כעת כתבו פונקציה המקבלת את מילון התוצאות ומחזירה מילון שבו שם המפלגה ממופה עם מספר המנדטים שקיבלה המפלגה.<br>
    הפעילו את פונקנציית <code>round</code> על תוצאת החלוקה (השתמשו ב־2,078) כדי לקבל את מספר המנדטים השלם.
</p>

<p style="text-align:right;direction:rtl;">
הדפיסו את המילון שקיבלתם. ודאו שסך המנדטים מסתכם ל־120.</p>

<p style="text-align:right;direction:rtl;">
כדי להרכיב קואליציה דרושים 61 מנדטים.<br>
המפלגות MidgeLandIsUs ו־NilsIsALeader לא מוכנות לשבת יחד בקואליציה.<br>
כתבו פונקציה המקבלת את המילון המכיל את תוצאות המנדטים ומדפיסה קואליציה אפשרית עבור NilsIsALeader וקואליציה אפשרית עבור MidgeLandIsUs.
</p> 

## <p style="text-align:right;direction:rtl;">Leveraging Simple Dictionaries</p>

### <p style="text-align:right;direction:rtl;">הקדמה</p>

<p style="text-align:right;direction:rtl;">
    <dfn>אוטומט סופי דטרמיניסטי</dfn> הוא מודל מתמטי שמגדיר אוסף של מצבים, וכללי מעבר ביניהם.<br>
    נפתח בדוגמה:<br>
    <figure>
        <img src="images/even_dfa.svg"  style="max-width:100%; margin-right: auto; margin-left: auto; text-align: center;" width="450px" alt="תמונת דוגמה לאוטומט סופי דטרמיניסטי. בצד ימין ישנו חץ שעליו כתוב 'התחלה'. החץ מצביע לעיגול שבו יש את המילים 'אי־זוגי', וממנו יוצא חץ לכיוון עיגול כפול שבו כתוב 'זוגי'. על החץ כתוב 'הספרה 2, 4, 6, 8 או 0'. מהעיגול הכפול יוצא חץ חזרה לעיגול הראשון, עליו כתוב 'הספרה 1, 3, 5, 7 או 9'. מהעיגול 'זוגי' יוצא חץ לעצמו עם המספרים 2, 4, 6, 8 ו־0. מהעיגול 'אי־זוגי' יוצא חץ לעצמו עם המספרים 1, 3, 5, 7, 9. ">
        <figcaption style="text-align: center; direction: rtl;">אוטומט סופי דטרמיניסטי שבודק האם מספר מסוים הוא מספר זוגי.</figcaption>
    </figure>
</p>

<p style="text-align:right;direction:rtl;">
    לפניכם מכונת מצבים שמקבלת מספר, ובודקת האם הוא זוגי.<br>
    כל מצב מיוצג בתמונה על ידי עיגול, וכללי המעבר ביניהם מיוצגים באמצעות חצים.
</p>

<p style="text-align:right;direction:rtl;">
    המכונה מתחילה מ"מצב התחלתי" נתון כלשהו, במקרה שלנו "מספר אי זוגי".<br>
    כל פעם שמתקבל קלט חדש, המכונה מחליטה לאיזה מצב לעבור לפי כללי המעבר הרשומים על החצים.<br>
    לדוגמה, עבור הקלט 2 ואז 7 ואז 4 (274) מכונת המצבים המופיעה מעלה תבצע את הפעולות הבאות:
</p>

<ol style="text-align: right; direction: rtl; float: right; clear: both;">
    <li>תכנס למצב ההתחלתי "מספר אי זוגי".</li>
    <li>תקרא את הספרה הראשונה במספר – 2, ותחליט לעבור (לפי החץ) למצב "מספר זוגי".</li> 
    <li>תקרא את הספרה השנייה במספר – 7, ותחליט לעבור (לפי החץ) למצב "מספר אי־זוגי".</li> 
    <li>תקרא את הספרה השלישית במספר – 4, ותחליט לעבור (לפי החץ) למצב "מספר זוגי".</li>
</ol>

<p style="text-align:right;direction:rtl;">
    לכל מצב באוטומט סופי דטרמיניסטי יש "סוג מצב": מקבל או לא מקבל.<br>
    מצב מקבל מסומן בעיגול כפול, כמו המצב "מספר זוגי" באוטומט דלעיל.<br>
    אם בסוף הריצה הגענו למצב מקבל, סימן שהקלט תקין, ואם לא – סימן שהקלט אינו תקין.<br>
    באוטומט הדוגמה שלנו כל קלט מספרי זוגי הוא תקין, וכל מספרי אי־זוגי נחשב ללא תקין.
</p>

<p style="text-align:right;direction:rtl;">
    כל אפשרות הכלולה במצבי המעבר נקראת <dfn>אות</dfn>. אוסף כל האותיות הכלולות באוטומט מסוים נקרא <dfn>א"ב הקלט</dfn>.<br>
    שימו לב שבאוטומט יכולה להיות כמות גדולה של מצבים וכללי־מעבר, ובכל אוטומט יכול להיות יותר ממצב מקבל אחד.<br>
    כמו כן, "אות" בא"ב יכולה לכלול יותר מתו אחד.
</p>

### <p style="text-align:right;direction:rtl;">תרגיל</p>

<p style="text-align:right;direction:rtl;">
   כתבו תוכנה שקוראת קובץ ויוצרת ממנו אוטומט סופי דטרמיניסטי.<br>
   בקובץ ישנו כל המידע הדרוש לבניית האוטומט, ורשימת קלטים. דוגמה לקובץ שכזה מופיעה מטה.<br>
   הדפיסו עבור כל קלט האם לאחר הרצתו האוטומט הגיע למצב מקבל או למצב לא מקבל.
</p>

<p style="text-align:right;direction:rtl;">
    הקובץ שהתוכנה שלכם אמורה לקרוא ונועד לתאר את האוטומט המופיע למעלה נראה כך:
</p>

<p style="text-align:right;direction:rtl;">
    על התוכנה שבניתם להדפיס במקרה הזה:
</p>

### <p style="text-align:right;direction:rtl;">הסבר עבור הדוגמה</p>

<p style="text-align:right;direction:rtl;">
    הפסקה הראשונה תמיד תראה כך:<br>
    הקובץ תמיד יפתח במילה <q dir="ltr">Definitions:</q>.<br>
    שורה אחריה תמיד תופיע המילה <q dir="ltr">States:</q>, ואחריה כל המצבים הקיימים במכונת המצבים, מופרדים בפסיק.<br>
    שורה אחריה תמיד תופיע המילה <q dir="ltr">Start:</q>, ואחריה המצב ההתחלתי של המכונה.<br>
    שורה אחריה תמיד תופיע המילה <q dir="ltr">Accepting:</q>, ואחריה כל המצבים שמוגדרים כמצבים מקבלים במכונת המצבים, מופרדים בפסיק.<br>
</p>

<p style="text-align:right;direction:rtl;">
    הפסקה השנייה תמיד תראה כך:<br>
    הפסקה תמיד תפתח במילה <q dir="ltr">Transitions:</q>, ואחריה רשימה של כללי מעבר.<br>
    כל כלל מעבר מורכב מהמצב ממנו יוצאים, המצב אליו מגיעים ומה הקלט ש"מפעיל" את המעבר.<br>
    השורה תכתב באופן הבא, כאשר X הוא המצב ממנו יוצאים, Z הוא המצב אליו מגיעים ו־Y היא רשימת הקלטים שמפעילה את מצב המעבר:<br>
    X: Y -> Z<br>
</p>

<p style="text-align:right;direction:rtl;">
    הפסקה השלישית תמיד תראה כך:<br>
    הפסקה תמיד תפתח במילה <q dir="ltr">Inputs:</q>, ואחריה רשימה של קלטים <strong>תקינים</strong>.<br>
    כל שורת קלט תופיע בשורה נפרדת בקובץ, ותכלול את רצף הקלטים שיועברו לאוטומט אחד אחרי השני, כאשר הם מופרדים בפסיק.
</p>

<p style="text-align:right;direction:rtl;">
    עבור הקלט הראשון, האוטומט יתחיל מהמצב ההתחלתי q0. האות 1 תשאיר אותו שם, כך גם 3, כך גם 1. מצב לא מקבל.<br>
    עבור הקלט השני, האוטומט יתחיל מהמצב ההתחלתי q0. האות 2 תעביר אותו ל־q1, האות 7 תעביר אותו ל־q0, האות 4 תעביר אותו חזרה ל־q1. מצב מקבל.<br>
    עבור הקלט השלישי, האוטומט יתחיל מהמצב ההתחלתי q0. האות 2 תעביר אותו ל־q1, האות 2 תשאיר אותו ב־q1 בכל המופעים הבאים שלה. מצב מקבל.<br>
</p>

In [28]:
#upload 68

def del_empty_spaces(data_list):

    i = 0
    while i < len(data_list):

        j = 0
        while j < len(data_list[i]):

            data_list[i][j] = data_list[i][j].strip()
            j += 1
        i += 1

    return data_list

def get_data_dict(data_list):

    print("im in get_data_dict")
    #  creating list of lists in order to prepare to make it a dictionary:
    index_of_input = data_list.index('Inputs:')

    i = 0
    while i < len(data_list):

        data_list[i] = data_list[i].split(":")
        i += 1
        
    print("data_list after splitting inner cells in ':'")

    print(data_list)
    print(f"data_list length: {len(data_list)}")

    data_dict = {}

    # deleting empty spaces:

    data_list = del_empty_spaces(data_list)

    i = 0
    while i < index_of_input:

        key_name = data_list[i][0]
        if key_name != 'Inputs':
            data_dict[data_list[i][0]] = data_list[i][1]
        i += 1
    
    print(data_dict)

    return data_dict

def get_data_list(data_str):

    print("Im in get_data_list")

    # creating a list out of the str:
    data_list = data_str.split("\n")

    # remove blank spaces from the list
    while "" in data_list:
        data_list.remove("")
    
    print("data_list:")
    print(data_list)

    print()

    return data_list

def get_index_input(data_list):

    print("im in get_index_input")
    index_of_input = data_list.index('Inputs:')

    return index_of_input

def get_data_from_file():

    print("im in get_data_from_file")
    with open('resources/automate-machine.txt' , 'r') as data_file:

        data_str = data_file.read()

        data_str = data_str.replace('->', ':')

        return data_str

def get_input_list(data_list,input_index):

    print("Im in get_input_list")
    input_list = []

    print(data_list)
    i = input_index + 1 
    while i < len(data_list):

        input_list.extend(data_list[i])
        i += 1

    i = 0
    while i < len(input_list):

        input_list[i] = input_list[i].split(',')
        i += 1

    input_list = del_empty_spaces(input_list)
    print(f"input list:")
    print(input_list)
    return input_list

def automate_machine(data_dict , input_list , transition_list):

    print("Im in automate_machine")
    accept_state = data_dict['Accepting']

    # run over the list of input
    i = 0
    while i < len(input_list):

        # run over the list of numbers of each input
        j = 0
        current_state = data_dict['Start']
        while j < len(input_list[i]):

            # run over the list of posible states and their transitions:
            current_num = input_list[i][j]
            r = 0
            while r < len(transition_list):

                state_to_go_out_from = transition_list[r][0]

                if current_state == state_to_go_out_from:

                    conditions_for_transition = transition_list[r][1]

                    if current_num in conditions_for_transition:

                        state_to_reach = transition_list[r][2]
                        current_state = state_to_reach
                        print(current_state)
                        # ending current loop, continuing to the next number in the list:
                        r = len(transition_list)
                r += 1    
            j += 1
        print(f"the result state for {input_list[i]} is : {current_state == accept_state}.")

        i += 1
  
def get_transition_list(data_list,input_index):

    print("im in get_transition_list ")
    transition_index = 0
    i = 0
    while i < len(data_list):

        if data_list[i][0] == 'Transitions':

            transition_index = i
        i += 1
    print(f"input index : {input_index}")

    i = transition_index + 1   
    transition_list = []
    while i < input_index:

        transition_list.append(data_list[i])
        i += 1

    print(f"trans list:   {transition_list}")

    return transition_list

def main():

    data_str = get_data_from_file()

    data_list = get_data_list(data_str)

    input_index = get_index_input(data_list)

    data_dict = get_data_dict(data_list)

    input_list = get_input_list(data_list,input_index)

    transition_list = get_transition_list(data_list , input_index)

    automate_machine(data_dict,input_list,transition_list)

main()

im in get_data_from_file
Im in get_data_list
data_list:
['Definitions:', 'States: q0, q1', 'Start: q0', 'Accepting: q1', 'Transitions:', 'q0: 2, 4, 6, 8, 0 : q1', 'q0: 1, 3, 5, 7, 9 : q0', 'q1: 2, 4, 6, 8, 0 : q1', 'q1: 1, 3, 5, 7, 9 : q0', 'Inputs:', '1, 3, 1', '2, 7, 4', '2, 2, 2, 2']

im in get_index_input
im in get_data_dict
data_list after splitting inner cells in ':'
[['Definitions', ''], ['States', ' q0, q1'], ['Start', ' q0'], ['Accepting', ' q1'], ['Transitions', ''], ['q0', ' 2, 4, 6, 8, 0 ', ' q1'], ['q0', ' 1, 3, 5, 7, 9 ', ' q0'], ['q1', ' 2, 4, 6, 8, 0 ', ' q1'], ['q1', ' 1, 3, 5, 7, 9 ', ' q0'], ['Inputs', ''], ['1, 3, 1'], ['2, 7, 4'], ['2, 2, 2, 2']]
data_list length: 13
{'Definitions': '', 'States': 'q0, q1', 'Start': 'q0', 'Accepting': 'q1', 'Transitions': '', 'q0': '1, 3, 5, 7, 9', 'q1': '1, 3, 5, 7, 9'}
Im in get_input_list
[['Definitions', ''], ['States', 'q0, q1'], ['Start', 'q0'], ['Accepting', 'q1'], ['Transitions', ''], ['q0', '2, 4, 6, 8, 0', 'q1'], ['q0', '1

### <p style="text-align:right;direction:rtl;">דוגמה נוספת</p>

<p style="text-align:right;direction:rtl;">
    זוהי דוגמה לקובץ שמתאר את האוטומט המופיע <a href="https://en.wikipedia.org/wiki/Deterministic_finite_automaton#/media/File:DFA_example_multiplies_of_3.svg">כאן</a>.
</p>

In [27]:
#upload 68

def del_empty_spaces(data_list):

    i = 0
    while i < len(data_list):

        j = 0
        while j < len(data_list[i]):

            data_list[i][j] = data_list[i][j].strip()
            j += 1
        i += 1

    return data_list

def get_data_dict(data_list):

    print("im in get_data_dict")
    #  creating list of lists in order to prepare to make it a dictionary:
    index_of_input = data_list.index('Inputs:')

    i = 0
    while i < len(data_list):

        data_list[i] = data_list[i].split(":")
        i += 1
    print(data_list)
    print(f"data_list length: {len(data_list)}")

    data_dict = {}

    # deleting empty spaces:

    data_list = del_empty_spaces(data_list)

    i = 0
    while i < index_of_input:

        key_name = data_list[i][0]
        if key_name != 'Inputs':
            data_dict[data_list[i][0]] = data_list[i][1]
        i += 1

    print(data_dict)

    return data_dict

def get_data_list(data_str):

    print("Im in get_data_list")

    # creating a list out of the str:
    data_list = data_str.split("\n")

    # remove blank spaces from the list
    while "" in data_list:
        data_list.remove("")

    print(data_list)

    print()

    return data_list

def get_index_input(data_list):

    print("im in get_index_input")
    index_of_input = data_list.index('Inputs:')

    return index_of_input

def get_data_from_file():

    print("im in get_data_from_file")
    with open('resources/automate-machine3.txt' , 'r') as data_file:

        data_str = data_file.read()

        data_str = data_str.replace('->', ':')

        return data_str

def get_input_list(data_list,input_index):

    print("Im in get_input_list")
    input_list = []

    print(data_list)
    i = input_index + 1 
    while i < len(data_list):

        input_list.extend(data_list[i])
        i += 1

    i = 0
    while i < len(input_list):

        input_list[i] = input_list[i].split(',')
        i += 1

    input_list = del_empty_spaces(input_list)
    print(f"input list:")
    print(input_list)
    return input_list

def automate_machine(data_dict , input_list , transition_list):

    print("Im in automate_machine")
    accept_state = data_dict['Accepting']

    # run over the list of input
    i = 0
    while i < len(input_list):

        # run over the list of numbers of each input
        j = 0
        current_state = data_dict['Start']
        while j < len(input_list[i]):

            # run over the list of posible states and their transitions:
            current_num = input_list[i][j]
            r = 0
            while r < len(transition_list):

                state_to_go_out_from = transition_list[r][0]

                if current_state == state_to_go_out_from:

                    conditions_for_transition = transition_list[r][1]

                    if current_num in conditions_for_transition:

                        state_to_reach = transition_list[r][2]
                        current_state = state_to_reach
                        print(current_state)
                        # ending current loop, continuing to the next number in the list:
                        r = len(transition_list)
                r += 1    
            j += 1
        print(f"the result state for {input_list[i]} is : {current_state == accept_state}.")

        i += 1
  
def get_transition_list(data_list,input_index):

    print("im in get_transition_list ")
    transition_index = 0
    i = 0
    while i < len(data_list):

        if data_list[i][0] == 'Transitions':

            transition_index = i
        i += 1
    print(f"input index : {input_index}")

    i = transition_index + 1   
    transition_list = []
    while i < input_index:

        transition_list.append(data_list[i])
        i += 1

    print(f"trans list:   {transition_list}")

    return transition_list

def main():

    data_str = get_data_from_file()

    data_list = get_data_list(data_str)

    input_index = get_index_input(data_list)

    data_dict = get_data_dict(data_list)

    input_list = get_input_list(data_list,input_index)

    transition_list = get_transition_list(data_list , input_index)

    automate_machine(data_dict,input_list,transition_list)

main()

im in get_data_from_file
Im in get_data_list
['Definitions:', 'States: q0, q1, q2', 'Start: q0', 'Accepting: q0', 'Transitions:', 'q0: 0, 3, 6, 9 : q0', 'q0: 2, 5, 8 : q1', 'q0: 1, 4, 7 : q2', 'q1: 1, 4, 7 : q0', 'q1: 0, 3, 6, 9 : q1', 'q1: 2, 5, 8 : q2', 'q2: 2, 5, 8 : q0', 'q2: 1, 4, 7 : q1', 'q2: 0, 3, 6, 9 : q2', 'Inputs:', '1', '1, 2', '1, 2, 3', '1, 3, 5', '1, 3, 8', '1, 2, 4, 1', '9, 9, 9, 9, 9, 9', '9, 9, 9, 9, 9, 1']

im in get_index_input
im in get_data_dict
[['Definitions', ''], ['States', ' q0, q1, q2'], ['Start', ' q0'], ['Accepting', ' q0'], ['Transitions', ''], ['q0', ' 0, 3, 6, 9 ', ' q0'], ['q0', ' 2, 5, 8 ', ' q1'], ['q0', ' 1, 4, 7 ', ' q2'], ['q1', ' 1, 4, 7 ', ' q0'], ['q1', ' 0, 3, 6, 9 ', ' q1'], ['q1', ' 2, 5, 8 ', ' q2'], ['q2', ' 2, 5, 8 ', ' q0'], ['q2', ' 1, 4, 7 ', ' q1'], ['q2', ' 0, 3, 6, 9 ', ' q2'], ['Inputs', ''], ['1'], ['1, 2'], ['1, 2, 3'], ['1, 3, 5'], ['1, 3, 8'], ['1, 2, 4, 1'], ['9, 9, 9, 9, 9, 9'], ['9, 9, 9, 9, 9, 1']]
data_list length: 23
{'D